// @aws-northstar/ui/components/CognitoAuth/components/ForgotPasswordView/index.js
import { jsx as _jsx } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import { useCallback, useMemo } from 'react';
import { componentTypes } from '../../../FormRenderer';
import GenericView from '../GenericView';
const ForgotPasswordView = ({ onResetPassword, onBackToSignIn, data }) => {
    const schema = useMemo(() => {
        const destination = data?.CodeDeliveryDetails?.Destination;
        const formSchema = {
            header: 'Reset Password',
            headerVariant: 'h2',
            variant: 'embedded',
            canCancel: false,
            submitLabel: 'Confirm',
            fields: [
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Code',
                    name: 'verificationCode',
                    placeholder: 'Enter verification code',
                    helperText: `A verification code has been sent to: ${destination}`,
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Password',
                    name: 'password',
                    type: 'password',
                    placeholder: 'Enter your Password',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Confirm Password',
                    name: 'confirmPassword',
                    type: 'password',
                    placeholder: 'Confirm your Password',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
            ],
        };
        return formSchema;
    }, [data]);
    const validate = useCallback((values) => {
        const errors = {};
        if (values.password !== values.confirmPassword) {
            errors.confirmPassword = 'Passwords do NOT match';
        }
        return errors;
    }, []);
    return (_jsx(GenericView, { schema: schema, onSubmit: onResetPassword, onBackToSignIn: onBackToSignIn, validate: validate }));
};
export default ForgotPasswordView;
