// @aws-northstar/ui/components/CognitoAuth/components/MFATotpSetupView/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import { useMemo } from 'react';
import { componentTypes } from '../../../FormRenderer';
import SpaceBetween from '@cloudscape-design/components/space-between';
import Box from '@cloudscape-design/components/box';
import Popover from '@cloudscape-design/components/popover';
import StatusIndicator from '@cloudscape-design/components/status-indicator';
import Button from '@cloudscape-design/components/button';
import QRCode from 'react-qr-code';
import { IMAGE_AUTH_APP } from '../../assets/images';
import GenericView from '../GenericView';
const MFATotpView = ({ secretCode, onConfirm, onBackToSignIn }) => {
    const CustomComponent = useMemo(() => {
        return () => {
            return (_jsxs(SpaceBetween, { size: "xl", direction: "horizontal", children: [_jsxs(SpaceBetween, { size: "xl", direction: "horizontal", children: [_jsx("span", { children: "1." }), _jsx("img", { width: 75, src: IMAGE_AUTH_APP, alt: "authenticator_app_mfa" }), _jsx("span", { children: "Install an authenticator app on your mobile device." })] }), _jsxs(SpaceBetween, { size: 'xl', direction: 'horizontal', children: [_jsx("span", { children: "2." }), _jsx(QRCode, { size: 75, value: `otpauth://totp/${window.location.origin}?secret=${secretCode}` }), _jsx("span", { style: { display: 'block', maxWidth: '320px' }, children: "Scan this QR code with your authenticator app. Alternatively, you can copy the secret key below and enter it in your authenticator app." }), _jsx(Box, { margin: "l", children: _jsx(Popover, { dismissButton: false, position: "top", size: "small", triggerType: "custom", content: _jsx(StatusIndicator, { type: "success", children: "Secret key copied" }), children: _jsx(Button, { iconName: "copy", onClick: () => navigator?.clipboard?.writeText?.(secretCode), children: "Copy secret key" }) }) })] })] }));
        };
    }, [secretCode]);
    const schema = useMemo(() => {
        return {
            header: 'Set Up Authenticator',
            headerVariant: 'h2',
            variant: 'embedded',
            canCancel: false,
            submitLabel: 'Continue',
            fields: [
                {
                    component: componentTypes.CUSTOM,
                    CustomComponent: CustomComponent,
                    name: 'description',
                },
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Code',
                    name: 'confirmationCode',
                    placeholder: 'Code',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
            ],
        };
    }, [CustomComponent]);
    return _jsx(GenericView, { schema: schema, onBackToSignIn: onBackToSignIn, onSubmit: onConfirm });
};
export default MFATotpView;
