// @cloudscape-design/board-components/test-utils/selectors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = exports.PaletteItemWrapper = exports.ItemsPaletteWrapper = exports.BoardItemWrapper = exports.BoardWrapper = exports.ElementWrapper = void 0;
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
Object.defineProperty(exports, "ElementWrapper", { enumerable: true, get: function () { return selectors_1.ElementWrapper; } });
const utils_1 = require("@cloudscape-design/test-utils-core/utils");
const index_1 = require("./board/index");
exports.BoardWrapper = index_1.default;
const index_2 = require("./board-item/index");
exports.BoardItemWrapper = index_2.default;
const index_3 = require("./items-palette/index");
exports.ItemsPaletteWrapper = index_3.default;
const index_4 = require("./palette-item/index");
exports.PaletteItemWrapper = index_4.default;
selectors_1.ElementWrapper.prototype.findBoard = function (selector) {
    const rootSelector = `.${index_1.default.rootSelector}`;
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findComponent(selector ? (0, utils_1.appendSelector)(selector, rootSelector) : rootSelector, index_1.default);
};
selectors_1.ElementWrapper.prototype.findBoardItem = function (selector) {
    const rootSelector = `.${index_2.default.rootSelector}`;
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findComponent(selector ? (0, utils_1.appendSelector)(selector, rootSelector) : rootSelector, index_2.default);
};
selectors_1.ElementWrapper.prototype.findItemsPalette = function (selector) {
    const rootSelector = `.${index_3.default.rootSelector}`;
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findComponent(selector ? (0, utils_1.appendSelector)(selector, rootSelector) : rootSelector, index_3.default);
};
selectors_1.ElementWrapper.prototype.findPaletteItem = function (selector) {
    const rootSelector = `.${index_4.default.rootSelector}`;
    // casting to 'any' is needed to avoid this issue with generics
    // https://github.com/microsoft/TypeScript/issues/29132
    return this.findComponent(selector ? (0, utils_1.appendSelector)(selector, rootSelector) : rootSelector, index_4.default);
};
var selectors_2 = require("@cloudscape-design/test-utils-core/selectors");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return selectors_2.createWrapper; } });
