// @aws-northstar/ui/components/FormRenderer/components/PlainText/index.js
import { jsx as _jsx } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import useFieldApi from '@data-driven-forms/react-form-renderer/use-field-api';
import TextContent from '@cloudscape-design/components/text-content';
const elementMapping = {
    h1: (label) => _jsx("h1", { children: label }),
    h2: (label) => _jsx("h2", { children: label }),
    h3: (label) => _jsx("h3", { children: label }),
    h4: (label) => _jsx("h4", { children: label }),
    h5: (label) => _jsx("h5", { children: label }),
    p: (label) => _jsx("p", { children: label }),
};
const PlainText = (props) => {
    const { label, element } = useFieldApi(props);
    return _jsx(TextContent, { children: element ? elementMapping[element]?.(label) || label : label });
};
export default PlainText;
