// @aws-northstar/ui/components/CognitoAuth/components/SignInView/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import SpaceBetween from '@cloudscape-design/components/space-between';
import Button from '@cloudscape-design/components/button';
import FormRenderer, { componentTypes } from '../../../FormRenderer';
import useSubmitCallback from '../../hooks/useSubmitCallback';
const schema = {
    variant: 'embedded',
    canCancel: false,
    submitLabel: 'Sign in',
    fields: [
        {
            component: componentTypes.TEXT_FIELD,
            isRequired: true,
            label: 'Username',
            name: 'Username',
            placeholder: 'Enter your Username',
            validate: [
                {
                    type: 'required',
                },
            ],
        },
        {
            component: componentTypes.TEXT_FIELD,
            isRequired: true,
            label: 'Password',
            name: 'Password',
            type: 'password',
            placeholder: 'Enter your Password',
            validate: [
                {
                    type: 'required',
                },
            ],
        },
    ],
};
const SignIn = ({ onSignIn, onForgotPassword }) => {
    const { handleSubmit, isSubmitting, errorMessage } = useSubmitCallback(onSignIn);
    return (_jsxs(SpaceBetween, { direction: "vertical", size: "xl", "data-testid": "sign-in-form", children: [_jsx(FormRenderer, { schema: schema, onSubmit: handleSubmit, errorText: errorMessage, isSubmitting: isSubmitting }), _jsx("div", { style: {
                    textAlign: 'center',
                }, children: _jsx(Button, { variant: "link", onClick: onForgotPassword, children: "Forgot your password?" }) })] }));
};
export default SignIn;
