// @cloudscape-design/board-components/test-utils/selectors/board-item/index.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const selectors_1 = require("@cloudscape-design/components/test-utils/selectors");
const selectors_2 = require("@cloudscape-design/test-utils-core/selectors");
const styles_selectors_js_1 = require("../../../board-item/styles.selectors.js");
const styles_selectors_js_2 = require("../../../internal/drag-handle/styles.selectors.js");
const styles_selectors_js_3 = require("../../../internal/resize-handle/styles.selectors.js");
class BoardItemWrapper extends selectors_2.ComponentWrapper {
    findDragHandle() {
        return this.findByClassName(styles_selectors_js_2.default.handle);
    }
    findResizeHandle() {
        return this.findByClassName(styles_selectors_js_3.default.handle);
    }
    findSettings() {
        return this.findByClassName(styles_selectors_js_1.default.settings);
    }
    // @cloudscape-design/components/container methods
    findHeader() {
        return this.findByClassName(styles_selectors_js_1.default["header-content"]);
    }
    findContent() {
        return (0, selectors_1.default)(this.getElement()).findContainer().findContent();
    }
    findFooter() {
        return (0, selectors_1.default)(this.getElement()).findContainer().findFooter();
    }
}
exports.default = BoardItemWrapper;
BoardItemWrapper.rootSelector = styles_selectors_js_1.default.root;
