// @aws-northstar/ui/components/CognitoAuth/components/Container/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import { useState, useEffect } from 'react';
import ContainerComponent from '@cloudscape-design/components/container';
import SpaceBetween from '@cloudscape-design/components/space-between';
import Header from '../Header';
const mediaMatch = window.matchMedia('(max-width: 600px)');
const styles = {
    container: (isMobileView) => isMobileView
        ? {}
        : {
            width: '100%',
            height: '100vh',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
        },
    content: (isMobileView) => isMobileView
        ? {}
        : {
            width: '480px',
        },
};
const Container = ({ children, header, logo }) => {
    const [matches, setMatches] = useState(mediaMatch.matches);
    useEffect(() => {
        const handler = (e) => setMatches(e.matches);
        mediaMatch.addEventListener('change', handler);
        return () => mediaMatch.removeEventListener('change', handler);
    });
    return (_jsx("div", { style: styles.container(matches), children: _jsx(ContainerComponent, { children: _jsxs(SpaceBetween, { direction: "vertical", size: "m", children: [(header || logo) && _jsx(Header, { logo: logo, children: header || '' }), _jsx("div", { style: styles.content(matches), children: children })] }) }) }));
};
export default Container;
