// @cloudscape-design/components/autosuggest/index.js
import { __rest } from "tslib";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import React from 'react';
import { getAnalyticsMetadataAttribute } from '@cloudscape-design/component-toolkit/internal/analytics-metadata';
import useBaseComponent from '../internal/hooks/use-base-component';
import { applyDisplayName } from '../internal/utils/apply-display-name';
import { getExternalProps } from '../internal/utils/external-props';
import InternalAutosuggest from './internal';
const Autosuggest = React.forwardRef((_a, ref) => {
    var { filteringType = 'auto', statusType = 'finished', disableBrowserAutocorrect = false } = _a, props = __rest(_a, ["filteringType", "statusType", "disableBrowserAutocorrect"]);
    const baseComponentProps = useBaseComponent('Autosuggest', {
        props: {
            autoFocus: props.autoFocus,
            disableBrowserAutocorrect,
            expandToViewport: props.expandToViewport,
            filteringType,
            readOnly: props.readOnly,
            virtualScroll: props.virtualScroll,
        },
    });
    const componentAnalyticsMetadata = {
        name: 'awsui.Autosuggest',
        label: 'input',
        properties: {
            disabled: `${!!props.disabled}`,
        },
    };
    const externalProps = getExternalProps(props);
    return (React.createElement(InternalAutosuggest, Object.assign({ filteringType: filteringType, statusType: statusType, disableBrowserAutocorrect: disableBrowserAutocorrect }, externalProps, baseComponentProps, { ref: ref }, getAnalyticsMetadataAttribute({ component: componentAnalyticsMetadata }))));
});
applyDisplayName(Autosuggest, 'Autosuggest');
export default Autosuggest;
