// @cloudscape-design/collection-hooks/cjs/use-collection.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCollection = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var react_1 = require("react");
var index_js_1 = require("./operations/index.js");
var utils_js_1 = require("./utils.js");
var use_collection_state_js_1 = require("./use-collection-state.js");
function useCollection(allItems, options) {
    var collectionRef = (0, react_1.useRef)(null);
    var _a = (0, use_collection_state_js_1.useCollectionState)(options, collectionRef), state = _a[0], actions = _a[1];
    var _b = (0, index_js_1.processItems)(allItems, state, options), items = _b.items, allPageItems = _b.allPageItems, pagesCount = _b.pagesCount, filteredItemsCount = _b.filteredItemsCount, actualPageIndex = _b.actualPageIndex, itemsTree = _b.itemsTree;
    var expandedItemsSet = new Set();
    if (options.expandableRows) {
        for (var _i = 0, _c = state.expandedItems; _i < _c.length; _i++) {
            var item = _c[_i];
            expandedItemsSet.add(options.expandableRows.getId(item));
        }
    }
    var visibleItems = items;
    if (options.expandableRows) {
        var flatItems_1 = new Array();
        var getId_1 = options.expandableRows.getId;
        var traverse_1 = function (items) {
            for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
                var item = items_1[_i];
                flatItems_1.push(item);
                if (expandedItemsSet.has(getId_1(item))) {
                    traverse_1(itemsTree.getChildren(item));
                }
            }
        };
        traverse_1(items);
        visibleItems = flatItems_1;
    }
    // Removing selected items that are no longer present in items unless keepSelection=true.
    if (options.selection && !options.selection.keepSelection) {
        var newSelectedItems = (0, index_js_1.processSelectedItems)(visibleItems, state.selectedItems, options.selection.trackBy);
        if (!(0, index_js_1.itemsAreEqual)(newSelectedItems, state.selectedItems, options.selection.trackBy)) {
            // This is a recommended pattern for how to handle the state, dependent on the incoming props
            // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
            actions.setSelectedItems(newSelectedItems);
        }
    }
    // Removing expanded items that are no longer present in items.
    if (options.expandableRows) {
        var newExpandedItems = visibleItems.filter(function (item) { return expandedItemsSet.has(options.expandableRows.getId(item)); });
        if (!(0, index_js_1.itemsAreEqual)(newExpandedItems, state.expandedItems, options.expandableRows.getId)) {
            actions.setExpandedItems(newExpandedItems);
        }
    }
    return __assign({ items: items, allPageItems: allPageItems, filteredItemsCount: filteredItemsCount, actions: actions }, (0, utils_js_1.createSyncProps)(options, state, actions, collectionRef, {
        actualPageIndex: actualPageIndex,
        pagesCount: pagesCount,
        allItems: allItems,
        allPageItems: allPageItems,
        itemsTree: itemsTree,
    }));
}
exports.useCollection = useCollection;
