// @data-driven-forms/react-form-renderer/form-renderer/render-form.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _setWith = _interopRequireDefault(require("lodash/setWith"));

var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));

var _reactFinalForm = require("react-final-form");

var _rendererContext = _interopRequireDefault(require("../renderer-context"));

var _condition = _interopRequireDefault(require("../condition"));

var _getConditionTriggers = _interopRequireDefault(require("../get-condition-triggers"));

var _prepareComponentProps = _interopRequireDefault(require("../prepare-component-props"));

var _excluded = ["component"],
    _excluded2 = ["condition", "hideField"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var FormFieldHideWrapper = function FormFieldHideWrapper(_ref) {
  var _ref$hideField = _ref.hideField,
      hideField = _ref$hideField === void 0 ? false : _ref$hideField,
      children = _ref.children;
  return hideField ? /*#__PURE__*/_react["default"].createElement("div", {
    hidden: true
  }, children) : children;
};

FormFieldHideWrapper.propTypes = {
  hideField: _propTypes["default"].bool,
  children: _propTypes["default"].oneOfType([_propTypes["default"].node, _propTypes["default"].arrayOf(_propTypes["default"].node)]).isRequired
};

var ConditionTriggerWrapper = function ConditionTriggerWrapper(_ref2) {
  var condition = _ref2.condition,
      values = _ref2.values,
      children = _ref2.children,
      field = _ref2.field;
  return /*#__PURE__*/_react["default"].createElement(_condition["default"], {
    condition: condition,
    values: values,
    field: field
  }, children);
};

ConditionTriggerWrapper.propTypes = {
  condition: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].array]),
  children: _propTypes["default"].node.isRequired,
  field: _propTypes["default"].object,
  values: _propTypes["default"].object.isRequired
};

var ConditionTriggerDetector = function ConditionTriggerDetector(_ref3) {
  var _ref3$values = _ref3.values,
      values = _ref3$values === void 0 ? {} : _ref3$values,
      _ref3$triggers = _ref3.triggers,
      triggers = _ref3$triggers === void 0 ? [] : _ref3$triggers,
      children = _ref3.children,
      condition = _ref3.condition,
      field = _ref3.field;
  var internalTriggers = (0, _toConsumableArray2["default"])(triggers);

  if (internalTriggers.length === 0) {
    return /*#__PURE__*/_react["default"].createElement(ConditionTriggerWrapper, {
      condition: condition,
      values: values,
      field: field
    }, children);
  }

  var name = internalTriggers.shift();
  return /*#__PURE__*/_react["default"].createElement(_reactFinalForm.Field, {
    name: name,
    subscription: {
      value: true,
      initial: true
    }
  }, function (_ref4) {
    var value = _ref4.input.value;
    return /*#__PURE__*/_react["default"].createElement(ConditionTriggerDetector, {
      triggers: (0, _toConsumableArray2["default"])(internalTriggers),
      values: (0, _setWith["default"])((0, _cloneDeep["default"])(values), name, value, Object),
      condition: condition,
      field: field
    }, children);
  });
};

ConditionTriggerDetector.propTypes = {
  values: _propTypes["default"].object,
  triggers: _propTypes["default"].arrayOf(_propTypes["default"].string),
  children: _propTypes["default"].node,
  condition: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].array]),
  field: _propTypes["default"].object.isRequired
};

var FormConditionWrapper = function FormConditionWrapper(_ref5) {
  var condition = _ref5.condition,
      children = _ref5.children,
      field = _ref5.field,
      conditionMapper = _ref5.conditionMapper;

  if (condition) {
    var triggers = (0, _getConditionTriggers["default"])(condition, field, conditionMapper);
    return /*#__PURE__*/_react["default"].createElement(ConditionTriggerDetector, {
      triggers: triggers,
      condition: condition,
      field: field
    }, children);
  }

  return children;
};

FormConditionWrapper.propTypes = {
  condition: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].array]),
  children: _propTypes["default"].node.isRequired,
  field: _propTypes["default"].object,
  conditionMapper: _propTypes["default"].object
};

var SingleField = function SingleField(_ref6) {
  var component = _ref6.component,
      rest = (0, _objectWithoutProperties2["default"])(_ref6, _excluded);

  var _useContext = (0, _react.useContext)(_rendererContext["default"]),
      actionMapper = _useContext.actionMapper,
      componentMapper = _useContext.componentMapper,
      conditionMapper = _useContext.conditionMapper;

  var _prepareComponentProp = (0, _prepareComponentProps["default"])({
    component: component,
    rest: rest,
    componentMapper: componentMapper,
    actionMapper: actionMapper
  }),
      componentProps = _prepareComponentProp.componentProps,
      Component = _prepareComponentProp.Component,
      overrideProps = _prepareComponentProp.overrideProps,
      mergedResolveProps = _prepareComponentProp.mergedResolveProps;

  var _componentProps$overr = _objectSpread(_objectSpread(_objectSpread({}, componentProps), overrideProps), mergedResolveProps && {
    resolveProps: mergedResolveProps
  }),
      condition = _componentProps$overr.condition,
      hideField = _componentProps$overr.hideField,
      restProps = (0, _objectWithoutProperties2["default"])(_componentProps$overr, _excluded2);

  return /*#__PURE__*/_react["default"].createElement(FormConditionWrapper, {
    condition: condition,
    field: restProps,
    conditionMapper: conditionMapper
  }, /*#__PURE__*/_react["default"].createElement(FormFieldHideWrapper, {
    hideField: hideField
  }, /*#__PURE__*/_react["default"].createElement(Component, restProps)));
};

SingleField.propTypes = {
  component: _propTypes["default"].string.isRequired,
  condition: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].array]),
  hideField: _propTypes["default"].bool,
  dataType: _propTypes["default"].string,
  validate: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].func, _propTypes["default"].object])),
  initialValue: _propTypes["default"].any,
  actions: _propTypes["default"].shape((0, _defineProperty2["default"])({}, _propTypes["default"].string, _propTypes["default"].func)),
  resolveProps: _propTypes["default"].func
};

var renderForm = function renderForm(fields) {
  return fields.map(function (field) {
    return Array.isArray(field) ? renderForm(field) : /*#__PURE__*/_react["default"].createElement(SingleField, (0, _extends2["default"])({
      key: field.name
    }, field));
  });
};

var _default = renderForm;
exports["default"] = _default;
