// @data-driven-forms/react-form-renderer/form-renderer/schema-error-component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var SchemaErrorComponent = function SchemaErrorComponent(_ref) {
  var name = _ref.name,
      message = _ref.message;
  return /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      padding: 5,
      border: '1px solid #bbb',
      borderCollapse: 'collapse',
      borderRadius: 2
    }
  }, /*#__PURE__*/_react["default"].createElement("h1", {
    style: {
      fontSize: 24,
      marginBottom: 5
    }
  }, "Form could not be rendered, because of invalid form schema."), /*#__PURE__*/_react["default"].createElement("h2", {
    style: {
      fontSize: 20,
      marginBottom: 5
    }
  }, name, ":"), /*#__PURE__*/_react["default"].createElement("p", {
    style: {
      marginBottom: 5
    }
  }, message), /*#__PURE__*/_react["default"].createElement("p", null, "If you don't know what this error means, please contact site administrator."));
};

SchemaErrorComponent.propTypes = {
  name: _propTypes["default"].string.isRequired,
  message: _propTypes["default"].string.isRequired
};
var _default = SchemaErrorComponent;
exports["default"] = _default;
