// @cloudscape-design/board-components/internal/dnd-controller/__mocks__/controller.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { useEffect } from "react";
import { vi } from "vitest";
import { EventEmitter } from "../event-emitter";
class MockController extends EventEmitter {
    start(event) {
        this.emit("start", event);
    }
    update(event) {
        this.emit("update", event);
    }
    submit() {
        this.emit("submit");
    }
    discard() {
        this.emit("discard");
    }
    acquire(event) {
        this.emit("acquire", event);
    }
}
export const mockController = new MockController();
export function useDragSubscription(event, handler) {
    useEffect(() => mockController.on(event, handler), [event, handler]);
}
export const mockDraggable = {
    start: vi.fn(),
    updateTransition: vi.fn(),
    submitTransition: vi.fn(),
    discardTransition: vi.fn(),
    getDroppables: vi.fn(),
};
export function useDraggable() {
    return mockDraggable;
}
export function useDroppable() { }
