// @aws-northstar/ui/components/CognitoAuth/components/MFASelectionView/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import { useMemo } from 'react';
import { componentTypes } from '../../../FormRenderer';
import SpaceBetween from '@cloudscape-design/components/space-between';
import Box from '@cloudscape-design/components/box';
import { IMAGE_AUTH_APP, IMAGE_SMS_MFA } from '../../assets/images';
import GenericView from '../GenericView';
const MFA_LITERALS = {
    SMS_MFA: {
        key: 'SMS',
        value: 'Authenticate using a code sent via text message to your phone.',
        img: IMAGE_SMS_MFA,
    },
    SOFTWARE_TOKEN_MFA: {
        key: 'Authenticator app',
        value: 'Authenticate using a code generated by an authenticator app.',
        img: IMAGE_AUTH_APP,
    },
};
const MFASelectionView = ({ challengeParams, onConfirm, onBackToSignIn }) => {
    const schema = useMemo(() => {
        return {
            header: 'Choose a MFA method',
            headerVariant: 'h2',
            variant: 'embedded',
            canCancel: false,
            submitLabel: 'Continue',
            fields: [
                {
                    component: componentTypes.PLAIN_TEXT,
                    label: 'You must configure MFA before you can sign in. Choose how you want to authenticate.',
                    name: 'description',
                },
                {
                    component: componentTypes.RADIO,
                    label: 'MFA Methods',
                    name: 'mfaMethod',
                    options: JSON.parse(challengeParams.MFAS_CAN_SETUP ?? challengeParams.MFAS_CAN_CHOOSE).map((v) => ({
                        value: v,
                        label: (_jsx(Box, { padding: {
                                bottom: 'l',
                            }, children: _jsxs(SpaceBetween, { direction: "horizontal", size: "xl", children: [_jsx("img", { src: MFA_LITERALS[v].img, alt: v }), _jsxs(SpaceBetween, { direction: "vertical", size: "xl", children: [_jsx("span", { children: _jsx("b", { children: MFA_LITERALS[v].key }) }), _jsx("span", { style: {
                                                    display: 'block',
                                                    wordWrap: 'break-word',
                                                    maxWidth: '360px',
                                                }, children: MFA_LITERALS[v].value })] })] }) })),
                    })),
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
            ],
        };
    }, [challengeParams]);
    return _jsx(GenericView, { schema: schema, onSubmit: onConfirm, onBackToSignIn: onBackToSignIn });
};
export default MFASelectionView;
