// @aws-northstar/ui/components/CognitoAuth/components/SignUpView/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/** *******************************************************************************************************************
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.                                                                              *
 ******************************************************************************************************************** */
import { useMemo } from 'react';
import FormRenderer, { componentTypes } from '../../../FormRenderer';
import Link from '@cloudscape-design/components/link';
import SpaceBetween from '@cloudscape-design/components/space-between';
import useSubmitCallback from '../../hooks/useSubmitCallback';
import AttributeMapping from '../../attributeMapping';
import validatePasswords from '../../utils/validatePasswords';
const SignUpView = ({ attributes, onSignUp, hrefTermsAndConditions }) => {
    const schema = useMemo(() => {
        const formSchema = {
            variant: 'embedded',
            canCancel: false,
            headerVariant: 'h2',
            submitLabel: 'Sign up',
            fields: [
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Username',
                    name: 'username',
                    placeholder: 'Enter your Username',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Password',
                    name: 'password',
                    type: 'password',
                    placeholder: 'Enter your Password',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
                {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: 'Confirm Password',
                    name: 'confirmPassword',
                    type: 'password',
                    placeholder: 'Confirm your Password',
                    validate: [
                        {
                            type: 'required',
                        },
                    ],
                },
            ],
        };
        attributes &&
            formSchema.fields.push(...attributes.map((attr) => {
                const displayName = attr.displayName || AttributeMapping[attr.name]?.displayName || attr.name;
                return {
                    component: componentTypes.TEXT_FIELD,
                    isRequired: true,
                    label: displayName,
                    placeholder: `Enter ${displayName}`,
                    name: `attributes[${attr.name}]`,
                    ...AttributeMapping[attr.name]?.componentSettingsOverride,
                    validate: [
                        ...(attr.required
                            ? [
                                {
                                    type: 'required',
                                },
                            ]
                            : []),
                        ...(AttributeMapping[attr.name]?.componentSettingsOverride?.validate || []),
                    ],
                };
            }));
        return formSchema;
    }, [attributes]);
    const { handleSubmit, isSubmitting, errorMessage } = useSubmitCallback(onSignUp);
    return (_jsxs(SpaceBetween, { direction: "vertical", size: "xl", "data-testid": "sign-up-form", children: [_jsx(FormRenderer, { schema: schema, onSubmit: handleSubmit, errorText: errorMessage, isSubmitting: isSubmitting, validate: validatePasswords }), hrefTermsAndConditions && (_jsx("div", { style: {
                    width: '100%',
                    textAlign: 'center',
                }, children: _jsx(Link, { href: hrefTermsAndConditions, external: true, children: "Terms and Conditions" }) }))] }));
};
export default SignUpView;
